File: mendeleev.f90
program mendeleev

   implicit none

   character(len=2), dimension(0:118), parameter :: ELEMENTS = (/ "? ", &
     "Ac", "Ag", "Al", "Am", "Ar", "As", "At", "Au", "B ", "Ba", "Be", "Bh", &
     "Bi", "Bk", "Br", "C ", "Ca", "Cd", "Ce", "Cf", "Cl", "Cm", "Cn", "Co", &
     "Cr", "Cs", "Cu", "Db", "Ds", "Dy", "Er", "Es", "Eu", "F ", "Fe", "Fl", &
     "Fm", "Fr", "Ga", "Gd", "Ge", "H ", "He", "Hf", "Hg", "Ho", "Hs", "I ", &
     "In", "Ir", "K ", "Kr", "La", "Li", "Lr", "Lu", "Lv", "Mc", "Md", "Mg", &
     "Mn", "Mo", "Mt", "N ", "Na", "Nb", "Nd", "Ne", "Nh", "Ni", "No", "Np", &
     "O ", "Og", "Os", "P ", "Pa", "Pb", "Pd", "Pm", "Po", "Pr", "Pt", "Pu", &
     "Ra", "Rb", "Re", "Rf", "Rg", "Rh", "Rn", "Ru", "S ", "Sb", "Sc", "Se", &
     "Sg", "Si", "Sm", "Sn", "Sr", "Ta", "Tb", "Tc", "Te", "Th", "Ti", "Tl", &
     "Tm", "Ts", "U ", "V ", "W ", "Xe", "Y ", "Yb", "Zn", "Zr" /)

   type :: element_t
      integer :: eid = 0
      character(len=:), pointer :: tail => null()
      type(element_t), pointer :: sibs => null()
      type(element_t), pointer :: next => null()
   end type element_t

   type(element_t), pointer :: root
   character(len=:), pointer :: word
   integer, dimension(:), allocatable :: formula
   integer :: length, argc, i

   argc = command_argument_count()
   do i = 1, argc
      call get_command_argument(i, length=length)
      allocate(character(len=length) :: word)
      call get_command_argument(i, value=word)
      write (*, "(A, ':')") word

      if (length > 0) then
         root => explode(word)
         allocate(formula(length))
         call print_plain(root, formula, 1)
         deallocate(formula)
         call free_elements(root)
      end if

      deallocate(word)
   end do


contains

   pure recursive subroutine free_elements(root)
      type(element_t), pointer, intent(in out) :: root
      type(element_t), pointer :: sibs

      do while (associated(root))
         if (associated(root%next)) call free_elements(root%next)
         sibs => root%sibs
         deallocate(root)
         root => sibs
      end do
   end subroutine free_elements


   pure integer function tolower(c)
      character(len=1), intent(in) :: c

      tolower = ior(32, iachar(c))
   end function tolower


   pure subroutine search(start, end, sh, c)
      integer, intent(in out) :: start, end
      integer, intent(in) :: sh
      character(len=1), intent(in) :: c
      integer :: l, m, u, c_

      c_ = tolower(c)

      u = end
      l = start
      do while (l < u)
         m = (u + l) / 2
         if (tolower(ELEMENTS(m)(sh:sh)) < c_) then
           l = m + 1
         else
           u = m
        endif
      end do

      if (l == end) then
        end = 0
        return
      end if

      if (tolower(ELEMENTS(l)(sh:sh)) /= c_) then
         end = 0
        return
      end if

      u = end
      start = l
      do while (l < u)
         m = (u + l) / 2
         if (c_ < tolower(ELEMENTS(m)(sh:sh))) then
           u = m
         else
           l = m + 1
        endif
      end do

      end = u
   end subroutine search


   function split(tail) result(head)
      character(len=:), pointer, intent(in) :: tail
      type(element_t), pointer :: head, last, el
      integer :: start, end, sh

      head => null()
      last => null()

      start = 1
      end = ubound(ELEMENTS, 1) + 1
      do sh = 1, len(tail)
         call search(start, end, sh, tail(sh:sh))
         if (start >= end) exit

         if (sh == len_trim(ELEMENTS(start))) then
            allocate(el)
            if (associated(last)) then
               last%sibs => el
            else
               head => el
            end if

            last => el
            last%eid = start
            last%tail => tail(sh+1:)

            start = start + 1
         end if
      end do

      if (.not. associated(head)) then
         allocate(head)
         head%tail => tail(2:)
      end if
   end function split


   recursive function explode(tail) result(root)
      character(len=:), pointer, intent(in) :: tail
      type(element_t), pointer :: root, el

      root => split(tail)
      el => root
      do while (associated(el))
         if (0 < len(el%tail)) el%next => explode(el%tail)
         el => el%sibs
      end do
   end function explode


   recursive subroutine print_plain(tree, formula, n)
      type(element_t), pointer, intent(in) :: tree
      integer, dimension(:), intent(in out) :: formula
      integer, value :: n
      type(element_t), pointer :: el
      integer :: i

      el => tree
      do while (associated(el))
         formula(n) = el%eid
         if (associated(el%next)) then
            call print_plain(el%next, formula, n+1)
         else
            do i = 1, n
               write (*, "(' ', A)", advance="no") trim(ELEMENTS(formula(i)))
            end do
            write (*, "()")
         end if
         el => el%sibs
      end do
   end subroutine print_plain

end program mendeleev

File: mendeleev.cpp
#include <cstddef>
#include <cstdlib>
#include <iostream>
#include <list>
#include <optional>
#include <string_view>
#include <utility>
#include <vector>

using std::list;
using std::optional;
using std::pair;
using std::string_view;
using std::vector;

static const vector<string_view> ELEMENTS = {
    "?",  "Ac", "Ag", "Al", "Am", "Ar", "As", "At", "Au", "B",  "Ba", "Be",
    "Bh", "Bi", "Bk", "Br", "C",  "Ca", "Cd", "Ce", "Cf", "Cl", "Cm", "Cn",
    "Co", "Cr", "Cs", "Cu", "Db", "Ds", "Dy", "Er", "Es", "Eu", "F",  "Fe",
    "Fl", "Fm", "Fr", "Ga", "Gd", "Ge", "H",  "He", "Hf", "Hg", "Ho", "Hs",
    "I",  "In", "Ir", "K",  "Kr", "La", "Li", "Lr", "Lu", "Lv", "Mc", "Md",
    "Mg", "Mn", "Mo", "Mt", "N",  "Na", "Nb", "Nd", "Ne", "Nh", "Ni", "No",
    "Np", "O",  "Og", "Os", "P",  "Pa", "Pb", "Pd", "Pm", "Po", "Pr", "Pt",
    "Pu", "Ra", "Rb", "Re", "Rf", "Rg", "Rh", "Rn", "Ru", "S",  "Sb", "Sc",
    "Se", "Sg", "Si", "Sm", "Sn", "Sr", "Ta", "Tb", "Tc", "Te", "Th", "Ti",
    "Tl", "Tm", "Ts", "U",  "V",  "W",  "Xe", "Y",  "Yb", "Zn", "Zr"};

typedef pair<size_t, string_view> Split;

struct Element {
  size_t eid;
  optional<list<Element>> next;
  Element(size_t eid_, optional<list<Element>> next_)
      : eid(eid_), next(next_) {}
};

static void search(size_t &start, size_t &end, size_t shift, char c) {
  size_t l, m, u;
  c |= ' ';

  u = end;
  l = start;
  while (l < u) {
    m = (l + u) / 2;
    if ((ELEMENTS[m][shift] | ' ') < c)
      l = m + 1;
    else
      u = m;
  }

  if ((l == end) || ((ELEMENTS[l][shift] | ' ') != c)) {
    end = 0;
    return;
  }

  u = end;
  start = l;
  while (l < u) {
    m = (l + u) / 2;
    if (c < (ELEMENTS[m][shift] | ' '))
      u = m;
    else
      l = m + 1;
  }

  end = u;
}

static list<Split> split(string_view tail) {
  auto x = list<Split>();

  size_t start = 1;
  size_t end = ELEMENTS.size();
  size_t shift = 0;

  while (shift < tail.length()) {
    search(start, end, shift, tail[shift]);
    if (start >= end)
      break;

    shift++;
    if (shift == ELEMENTS[start].length()) {
      x.emplace_back(start, tail.substr(shift));
      start++;
    }
  }

  if (x.empty())
    x.emplace_back(0, tail.substr(1));

  return x;
}

static list<Element> explode(string_view tail) {
  auto x = list<Element>();
  for (auto const &s : split(tail)) {
    x.emplace_back(s.first, s.second.empty()
                                ? std::nullopt
                                : std::make_optional(explode(s.second)));
  }
  return x;
}

static void print_plain(list<Element> const &tree, vector<size_t> &formula) {
  for (auto const &x : tree) {
    formula.push_back(x.eid);
    if (x.next)
      print_plain(*x.next, formula);
    else {
      for (auto i : formula)
        std::cout << " " << ELEMENTS[i];
      std::cout << std::endl;
    }
    formula.pop_back();
  }
}

int main(int argc, const char *argv[]) {
  vector<size_t> formula;

  for (int i = 1; i < argc; i++) {
    string_view word = argv[i];
    std::cout << word << ":" << std::endl;
    if (auto len = word.length(); len) {
      formula.reserve(len);
      print_plain(explode(word), formula);
    }
  }

  return EXIT_SUCCESS;
}
